### ISSUE TITLE
CodeActions are very very chatty

### ISSUE THREAD

**Issue by alexandrudima** [Microsoft/vscode#57875]
Run with `--logExtensionHostCommunication` or with code change for `LOG_EXTENSION_HOST_COMMUNICATION`

* Create an empty folder
* Create a javascript file
* Enable auto save
* Type one character in the javascript file

![image](https://user-images.githubusercontent.com/5047891/45038813-4ecad780-b062-11e8-8af7-8ebaa3ab38d6.png)

Obseve, the renderer process asks **9 times** for code actions after typing one character.

<img width="1362" alt="screen shot 2018-09-04 at 16 48 33" src="https://user-images.githubusercontent.com/5047891/45038882-7883fe80-b062-11e8-9a7b-09efc006fd58.png">

**Comment by mjbvz** [comment]
The three `$provideCodeActions` requests per group are:

* JS/TS quick fixes
* JS/TS refactors
* JS/TS source actions

A few thoughts on fixing this:

1. Use a single message to the extension host to get code actions from all providers instead of one message per provider. Not sure if any of our other language features also use this mode of communications. @alexandrudima / @jrieken Any thoughts here?

1. Avoid calling code action providers that we know (from their metadata) will not return actions we are interested in. In the js/ts case, we could avoid calling the source action provider. 

---

The three groups of requests are caused by:

- Cursor change
- Markers change on document change
- Markers change on document save

To address this:

1. Don't update markers if they have not really changed. We could try handling this in the js/ts extension, in the extension host, or in the main thread. We can address the basic case with zero markers first (if we have not already done so)

1. Debounce the actual code action trigger. We currently debounce the cursor and marker events separately, which means that both can still be triggered immediately after one another.

**Comment by mjbvz** [comment]
Addressed `$provideCodeActions` part 2 with  72a24483c490383fc6caf3a412d01d92ef2a0f99. This eliminates 3 of the requests. We still have two separate calls for the refactor and quick fix providers

**Comment by jrieken** [comment]
> Use a single message to the extension host to get code actions from all providers instead of one message per provider.

We have only done that for decoration providers (the scm and marker decorations) but no language yet. I think that something to optimise we really have a problem here

**Comment by jrieken** [comment]
I have pushed a change to prevent marker changes when nothing really has changed (on the extension host side)

**Comment by jrieken** [comment]
<img width="982" alt="screen shot 2018-09-11 at 12 47 40" src="https://user-images.githubusercontent.com/1794099/45355521-e4b1b580-b5c0-11e8-9272-ed1952e5d059.png">

We are down to two calls - one per provider which I think is fair

**Comment by alexandrudima** [comment]
:+1: Thank you, this is cool!

Is there any limitation why we can't have TS registering a single provider and then joining the results in the extension side?

**Comment by mjbvz** [comment]
Each ts code action provider has its own metadata that indicates what type of code actions it can return. This let's us avoid calling one or the other in cases like using the refactoring keybinding

**Comment by alexandrudima** [comment]
> This let's us avoid calling one or the other in cases like using the refactoring keybinding

We invoke two times when typing a letter in an editor but we are able to **optimize** and invoke only once when using the refactoring keybinding ?

But... why are we making worse the 99% case (typing) vs the 1% case (invoking the refactoring keybinding)

**Comment by jrieken** [comment]
I think that design was driven by how expensive it is to compute refactorings or quick fixes. E.g. we know that the refactor-menu won't show quick fixes so why ask quick-fix-only code action providers for actions?

**Comment by alexandrudima** [comment]
I think I understand this better after chatting with @jrieken ... it's ok with me to leave these calls in given the current design.
